### ISSUE TITLE
How does this compile and exclude the C-Runtime library without specifying the /nodefaultlib option?

### ISSUE THREAD

**Issue by cmdusr** [cmuratori/refterm#37]
Given that:

    1. build.bat does not use the /nodefaultlib option
    2. /nodefaultlib seems to kill the #pragma comment(lib, "libname") directives which are used in refterm.c.
        https://stackoverflow.com/questions/43131487/pragma-commentlib-not-working

How does this exclude the C Runtime?
Is the option excluded so that #pragma comment(lib, "libname") can be used?

**Comment by mmozeiko** [comment]
Because of `WinMainCRTStartup` entry point and `/opt:ref` linker argument.

The way linker works is following - it starts with this function, finds all functions called from it and takes their code, then repeats process for those functions until every function is put into output file. If you don't reference/call any C runtime function, no extra code from C runtime will be put in output file. And `/opt:ref` argument means that any non referenced function will be eliminated from output file.

If you call this function just `main` (or `WinMain`) then linker will take large part of C runtime in output executable, because C runtime is the one that provides mainCRTStartup/WinMainCRTStartup which prepares C runtime and then calls your main/WinMain - and this code around uses a lot of C runtime functions.

Basically mainCRTStartup & WinMainCRTStartup functions are special in a way that if you have them the linker starts takes their code, but if you don't have them then linker uses C runtime lib that provides them (and expects main/WinMain in your code).

**Comment by cmdusr** [comment]
Thank you for your detailed explanation! That answers my main question regarding the C runtime library.

I am assuming the reason to use `/nodefaultlib` is so that each library must be explicitly listed as arguments for the compiler in the makefile / build.bat.

Why is using `/opt:ref` used instead? 
Personal preference? Convivence?

**Comment by mmozeiko** [comment]
I like to avoid /nodefaultlib because that prevents using #pragma comment(lib, ...) in source.
Also this way it allows you to use some small parts of C runtime if you really needed, like calling `sinf` or similar.

That said /nodefaultlib is useful if you want to check & guarantee that your code does not call anything from C runtime.

**Comment by cmuratori** [comment]
I didn't actually realize, before looking at Mārtiņš' original prototype, that any of this worked, BTW. It's actually really great to know, and I will probably switch to doing things this way in future projects.

The primary motivation from my perspective is that you can remove the CRT but still list all the libraries in the C file, which means you do not have to list them on the cc line.  This might _seem_ irrelevant, but actually, if you build with multiple compilers (as here), then this saves you the trouble of listing them twice.  CLANG and MSVC, for example, have two different ways of listing the libraries on the command line (one uses spaces, the other commas), so putting them in the C file makes it much easier, and avoids resorting to .bat list processing heroics or figuring out better command line switches, etc.

So I think this is a really nice trick!

\- Casey

**Comment by cmdusr** [comment]
Thanks! That answers everything.

Note: If using `/opt:ref` in your own project make sure to mark the crt entrypoint (`WinMainCRTStartup` or `mainCRTStartup`) as `extern 'C'` in your .cpp file, otherwise the C runtime library will be included.
It works in this case because refterm is a .c file.
